File: lua_tests/in_fov_test.lua
--[[ Copyright (C) 2018-2019 Google Inc.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
]]

local asserts = require 'testing.asserts'
local test_runner = require 'testing.test_runner'
local helpers = require 'common.helpers'
local game = require 'dmlab.system.game'

local tests = {}

local ORIGIN = {0, 0, 0}
local FORWARD = {100, 0, 0}
local BACK = {-100, 0, 0}
local LEFT = {0, 100, 0}
local RIGHT = {0, -100, 0}
local DOWN = {0, 0, -100}
local UP = {0, 0, 100}

local NULL_ROT = {0, 0, 0}
local YAW_180 = {0, 180, 0}
local YAW_LEFT_89 = {0, 89, 0}
local YAW_LEFT_90 = {0, 90, 0}
local YAW_LEFT_91 = {0, 91, 0}
local YAW_RIGHT_90 = {0, -90, 0}
local PITCH_DOWN_90 = {-90, 0, 0}
local PITCH_UP_90 = {90, 0, 0}

function tests:narrowFovForward()
  assert(game:inFov(ORIGIN, FORWARD, NULL_ROT, 1))
end

function tests:narrowFovLeft()
  assert(game:inFov(ORIGIN, LEFT, YAW_LEFT_90, 1))
end

function tests:narrowFovRight()
  assert(game:inFov(ORIGIN, RIGHT, YAW_RIGHT_90, 1))
end

function tests:narrowFovBack()
  assert(game:inFov(ORIGIN, BACK, YAW_180, 1))
end

function tests:narrowFovDown()
  assert(game:inFov(ORIGIN, DOWN, PITCH_DOWN_90, 1))
end

function tests:narrowFovUp()
  assert(game:inFov(ORIGIN, UP, PITCH_UP_90, 1))
end

function tests:nullFov()
  asserts.EQ(false, game:inFov(ORIGIN, FORWARD, NULL_ROT, 0))
end

function tests:edgeIn()
  asserts.EQ(true, game:inFov(ORIGIN, {110, 100, 100}, NULL_ROT, 90))
end

function tests:edgeOut()
  asserts.EQ(false, game:inFov(ORIGIN, {90, 100, 100}, NULL_ROT, 90))
end

function tests:edgeInRot()
  asserts.EQ(true, game:inFov(ORIGIN, {100, 110, 100}, YAW_LEFT_90, 90))
end

function tests:edgeOutRot()
  asserts.EQ(false, game:inFov(ORIGIN, {100, 90, 100}, YAW_LEFT_90, 90))
end

function tests:fovIn()
  asserts.EQ(true, game:inFov(ORIGIN, {100, 100, 100}, NULL_ROT, 91))
end

function tests:fovOut()
  asserts.EQ(false, game:inFov(ORIGIN, {100, 100, 100}, NULL_ROT, 89))
end

function tests:fovInRot()
  asserts.EQ(true, game:inFov(ORIGIN, {100, 100, 100}, YAW_LEFT_90, 91))
end

function tests:fovOutRot()
  asserts.EQ(false, game:inFov(ORIGIN, {100, 100, 100}, YAW_LEFT_90, 89))
end

function tests:angleIn()
  asserts.EQ(true, game:inFov(ORIGIN, {100, 100, 100}, YAW_LEFT_89, 90))
end

function tests:angleOut()
  asserts.EQ(false, game:inFov(ORIGIN, {100, 100, 100}, YAW_LEFT_91, 90))
end

return test_runner.run(tests)

File: python/tests/in_fov_test.py
# Copyright 2017-2018 Google Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
"""Tests for InFov observations."""

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import os
from absl.testing import absltest
import numpy as np
import six

import deepmind_lab


class InFovTest(absltest.TestCase):

  def test_pickup_raycast(self):
    env = deepmind_lab.Lab(
        'tests/in_fov_test', ['FOVTESTS', 'ENTITYVIS', 'DEBUG.POS.TRANS'],
        config={
            'fps': '60',
            'width': '80',
            'height': '80'
        })

    action_spec = env.action_spec()
    action_index = {action['name']: i for i, action in enumerate(action_spec)}

    action = np.zeros([len(action_spec)], dtype=np.intc)

    action[action_index['MOVE_BACK_FORWARD']] = 1

    reward = env.reset()

    fov_tests = env.observations()['FOVTESTS']
    self.assertEqual(fov_tests[0], 1)
    self.assertEqual(fov_tests[1], 1)
    self.assertEqual(fov_tests[2], 0)
    entity_vis = env.observations()['ENTITYVIS']
    self.assertEqual(entity_vis[0], 1)
    self.assertLess(entity_vis[1], 1)
    self.assertEqual(entity_vis[2], -1)

    for _ in six.moves.range(120):
      reward = env.step(action, 1)
      if reward > 0:
        self.assertEqual(reward, 1)
        break
    else:
      print(env.observations()['DEBUG.POS.TRANS'])
      self.fail('Failed to pickup 1st apple!')

    fov_tests = env.observations()['FOVTESTS']
    self.assertEqual(fov_tests[0], 1)
    self.assertEqual(fov_tests[1], 0)
    self.assertEqual(fov_tests[2], 0)
    entity_vis = env.observations()['ENTITYVIS']
    self.assertEqual(entity_vis[0], 1)
    self.assertEqual(entity_vis[1], -1)
    self.assertEqual(entity_vis[2], -1)

    action[action_index['STRAFE_LEFT_RIGHT']] = 1
    action[action_index['MOVE_BACK_FORWARD']] = 0

    for _ in six.moves.range(160):
      reward = env.step(action, 1)
      if reward > 0:
        self.assertEqual(reward, 1)
        break
    else:
      print(env.observations()['DEBUG.POS.TRANS'])
      self.fail('Failed to pickup 2st apple!')

    fov_tests = env.observations()['FOVTESTS']
    self.assertEqual(fov_tests[0], 0)
    self.assertEqual(fov_tests[1], 0)
    self.assertEqual(fov_tests[2], 0)
    entity_vis = env.observations()['ENTITYVIS']
    self.assertEqual(entity_vis[0], -1)
    self.assertEqual(entity_vis[1], -1)
    self.assertEqual(entity_vis[2], -1)

    action[action_index['STRAFE_LEFT_RIGHT']] = 0
    action[action_index['MOVE_BACK_FORWARD']] = -1

    for _ in six.moves.range(160):
      reward = env.step(action, 1)
      if reward > 0:
        self.assertEqual(reward, 1)
        break
    else:
      print(env.observations()['DEBUG.POS.TRANS'])
      self.fail('Failed to pickup 3rd apple!')

    fov_tests = env.observations()['FOVTESTS']
    self.assertEqual(fov_tests[0], 1)
    self.assertEqual(fov_tests[1], 1)
    self.assertEqual(fov_tests[2], 0)

if __name__ == '__main__':
  if 'TEST_SRCDIR' in os.environ:
    deepmind_lab.set_runfiles_path(
        os.path.join(os.environ['TEST_SRCDIR'],
                     'org_deepmind_lab'))
  absltest.main()
